modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/CalciteQueryProcessor.java

```diff
@@ -326,89 +327,114 @@ public PrepareServiceImpl prepareService() {
+                || qryNode.getKind() == SqlKind.MERGE || qryNode.getKind() == SqlKind.DELETE;
 
-            try {
-                if (qryList.size() == 1) {
-                    plan = queryPlanCache().queryPlan(
-                        new CacheKey(schemaName, sql), // Use source SQL to avoid redundant parsing next time.
-                        () -> prepareSvc.prepareSingle(sqlNode, qry.planningContext()));
-                }
-                else
-                    plan = prepareSvc.prepareSingle(sqlNode, qry.planningContext());
+        List<FieldsQueryCursor<List<?>>> cursors = new ArrayList<>(batchedParams.size());
 
-                cursors.add(executionSvc.executePlan(qry, plan));
-            }
-            catch (Exception e) {
-                boolean isCanceled = qry.isCancelled();
+        List<RootQuery<Object[]>> qrys = new ArrayList<>(batchedParams.size());
 
-                qrys.forEach(RootQuery::cancel);
+        PlanSupplier<Object[]> planSupplier = new PlanSupplier<Object[]>() {
+            private QueryPlan plan;
 
-                qryReg.unregister(qry.id());
+            @Override public QueryPlan get(RootQuery<Object[]> qry) {
+                if (plan == null)
+                    plan = prepareSvc.prepareSingle(qryNode, qry.planningContext());
 
-                if (isCanceled)
-                    throw new IgniteSQLException("The query was cancelled while planning", IgniteQueryErrorCode.QUERY_CANCELED, e);
-                else
-                    throw e;
+                return plan;
             }
-        }
+        };
+
+        for (final Object[] batch: batchedParams)
+            cursors.add(executeQuery(qryCtx, planSupplier, schemaName, qryNode.toString(), qrys, batch));
 
         return cursors;
     }
 
+    /** */
+    private FieldsQueryCursor<List<?>> executeQuery(
+        @Nullable QueryContext qryCtx,
+        @Nullable PlanSupplier<Object[]> plan,
+        String schema,
+        String sql,
+        @Nullable List<RootQuery<Object[]>> qrys,
+        Object... params
+    ) {
+        RootQuery<Object[]> qry = new RootQuery<>(
+            sql,
+            schemaHolder.schema(schema),
+            params,
+            qryCtx,
+            exchangeSvc,
+            (q) -> qryReg.unregister(q.id()),
+            log,
+            queryPlannerTimeout
+        );
+
+        if (qrys != null)
+            qrys.add(qry);
+
+        qryReg.register(qry);
+
+        try {
+            return executionSvc.executePlan(qry, plan.get(qry));
+        }
+        catch (Exception e) {
+            boolean isCanceled = qry.isCancelled();
+
+            if (qrys != null)
+                qrys.forEach(RootQuery::cancel);
+
+            qryReg.unregister(qry.id());
+
+            if (isCanceled)
+                throw new IgniteSQLException("The query was cancelled while executing", IgniteQueryErrorCode.QUERY_CANCELED, e);
```

Comment:
Let's keep error message "while planning", since execution is async and most likely we catch error in sync `plan.get()` method, or in mapping phase of `executePlan` method (which is rather related to planning than executing)